src/sys/unix/mod.rs

```diff
@@ -118,6 +118,53 @@ impl Socket {
         }
     }
 
+    pub fn connect_timeout(&self, addr: &SocketAddr, timeout: Duration) -> io::Result<()> {
+        self.set_nonblocking(true)?;
+
+        match self.connect(addr) {
+            Ok(()) => {
+                self.set_nonblocking(false)?;
+                return Ok(());
+            }
+            // there's no io::ErrorKind conversion registered for EINPROGRESS :(
+            Err(ref e) if e.raw_os_error() == Some(libc::EINPROGRESS) => {}
+            Err(e) => return Err(e),
+        }
+
+        let mut pollfd = libc::pollfd {
+            fd: self.as_raw_fd(),
+            events: libc::POLLOUT,
+            revents: 0,
+        };
+
+        if timeout.as_secs() == 0 && timeout.subsec_nanos() == 0 {
+            return Err(io::Error::new(io::ErrorKind::InvalidInput,
+                                    "cannot set a 0 duration timeout"));
+        }
+
+        let mut timeout = timeout.as_secs()
+            .saturating_mul(1_000)
+            .saturating_add(timeout.subsec_nanos() as u64 / 1_000_000);
+        if timeout == 0 {
+            timeout = 1;
+        }
+
+        let timeout = cmp::min(timeout, c_int::max_value() as u64) as c_int;
+
+        let r = unsafe { libc::poll(&mut pollfd, 1, timeout) };
```

Comment:
What should we do about the timeout in that situation? Just use the original one, or try to keep it updated with a couple of Instants?

Doesn't everything use SA_RESTART, anyway?